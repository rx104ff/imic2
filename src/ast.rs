// src/ast.rs
#[derive(Debug, Clone)]
pub enum Expr {
    Number(i64),
    Var(String),
    Let(String, Box<Expr>, Box<Expr>),
    If(Box<Expr>, Box<Expr>, Box<Expr>),
    Lambda(String, Box<Expr>),
    App(Box<Expr>, Box<Expr>),
    BinOp(String, Box<Expr>, Box<Expr>),
}
