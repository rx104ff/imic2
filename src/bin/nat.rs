use std::env;
use imic2::{common, nat};

fn main() {
    let args: Vec<String> = env::args().collect();
    let (mode, input_str) = match args.len() {
        // Case 1: No mode provided (e.g., `cargo run --bin nat "judgment"`)
        2 => (None, &args[1]),
        // Case 2: Mode provided (e.g., `cargo run --bin nat CompareNat1 "judgment"`)
        3 => (Some(&args[1]), &args[2]),
        _ => {
            eprintln!("Usage: cargo run --bin nat [<MODE>] \"<JUDGMENT>\"");
            eprintln!("Default mode handles 'plus' and 'times'.");
            eprintln!("Optional modes: CompareNat1, CompareNat2, CompareNat3");
            return;
        }
    };

    // Parse the optional mode string into our enum
    let comparison_mode: Option<nat::ast::ComparisonMode> = match mode {
        None => None,
        Some(s) => match s.as_str() {
            "CompareNat1" => Some(nat::ast::ComparisonMode::V1),
            "CompareNat2" => Some(nat::ast::ComparisonMode::V2),
            "CompareNat3" => Some(nat::ast::ComparisonMode::V3),
            _ => {
                eprintln!("Error: Unknown comparison mode '{}'", s);
                return;
            }
        },
    };

    let tokens = common::token::tokenize(input_str);
    let mut parser = nat::parser::Parser::new(tokens);
    let judgment = match parser.parse_judgment() {
        Ok(j) => j,
        Err(e) => { eprintln!("Parsing Error: {}", e); return; }
    };

    // Call the derivator with the optional mode
    match nat::nat::derive(&judgment, comparison_mode) {
        Ok(derivation) => println!("{}", derivation),
        Err(e) => eprintln!("Derivation Error: {}", e),
    }
}
