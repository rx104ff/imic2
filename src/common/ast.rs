use std::fmt;
use std::collections::{HashSet};
use std::hash::{Hash, Hasher};
use std::cmp::Ordering;

// --- Universal Primitives ---

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct Var(pub String);

#[derive(Debug, Clone, PartialEq)]
pub enum Op { Add, Sub, Mul, Lt, Cons }

// --- Types for Nat Language ---

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Nat {
    Z,
    S(Box<Nat>),
}

// --- Types for Type Systems (TypingML4 & PolyTypingML4) ---

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Type {
    Int,
    Bool,
    Fun(Box<Type>, Box<Type>),
    List(Box<Type>),
    Var(TypeVar),
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct TyScheme {
    pub vars: Vec<TypeVar>,
    pub ty: Type,
}

#[derive(Debug, Clone)]
pub struct TypeVar {
    pub id: usize,
    pub name: String,
}

impl PartialEq for TypeVar { 
    fn eq(&self, other: &Self) -> bool { 
        self.id == other.id 
    } 
}

impl Eq for TypeVar {}

impl PartialOrd for TypeVar { 
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> { 
        Some(self.cmp(other)) 
    } 
}

impl Ord for TypeVar { 
    fn cmp(&self, other: &Self) -> Ordering { 
        self.id.cmp(&other.id) 
    } 
}

impl Hash for TypeVar { 
    fn hash<H: Hasher>(&self, state: &mut H) { 
        self.id.hash(state); 
    } 
}

pub type TypeEnv = Vec<(Var, TyScheme)>;

// --- Universal Expression AST ---

#[derive(Debug, Clone, PartialEq)]
pub enum Expr {
    // Literals for ML, Nat, and Type systems
    Int(i64),
    Bool(bool),
    Nat(Nat),
    Var(Var),
    Nil,
    
    // ML & Type System Expressions
    Let(Var, Box<Expr>, Box<Expr>, bool),
    LetRec(Var, Var, Box<Expr>, Box<Expr>, bool),
    Fun(Var, Box<Expr>, bool),
    App(Box<Expr>, Box<Expr>, bool),
    If(Box<Expr>, Box<Expr>, Box<Expr>, bool),
    BinOp(Box<Expr>, Op, Box<Expr>, bool),
    Match(Box<Expr>, Box<Expr>, Var, Var, Box<Expr>, bool),
    
    // Nat Expression extensions
    Plus(Box<Expr>, Box<Expr>),
    Times(Box<Expr>, Box<Expr>),
}

// --- Universal Judgment AST ---

// The Judgment enum can now represent a judgment from ANY of your language systems.
#[derive(Debug, Clone, PartialEq)]
pub enum Judgment {
    // For ML evaluation
    EvaluatesTo(Expr, Type), // Assuming Type can also represent ML values
    
    // For Nat arithmetic
    Is(Expr, Type), // e.g., `1 plus 2 is 3`
    
    // For Nat reduction
    ReducesTo(Expr, Expr, String), // e.g., `e1 ---> e2`
    
    // For Type Checking
    Infer(TypeEnv, Expr, Type),
    
    // For Polymorphic Inference
    PolyInfer(TypeEnv, Expr, Type),
}

// --- All Display and Helper Implementations ---
impl fmt::Display for Nat {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Nat::Z => write!(f, "Z"),
            Nat::S(n) => write!(f, "S({})", n),
        }
    }
}

impl fmt::Display for Var {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.0)
    }
}

impl fmt::Display for Op {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Op::Add => write!(f, "+"), Op::Sub => write!(f, "-"), Op::Mul => write!(f, "*"),
            Op::Lt => write!(f, "<"), Op::Cons => write!(f, "::"),
        }
    }
}

impl fmt::Display for Expr {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Expr::Nat(n) => write!(f, "{}", n),
            Expr::Plus(e1, e2) => write!(f, "({} + {})", e1, e2),
            Expr::Times(e1, e2) => write!(f, "({} * {})", e1, e2),
            Expr::Int(n) => write!(f, "{}", n),
            Expr::Bool(b) => write!(f, "{}", b),
            Expr::Var(v) => write!(f, "{}", v.0),
            Expr::Nil => write!(f, "[]"),
            Expr::Let(x, e1, e2, is_paren) => {
                let s = format!("let {} = {} in {}", x, e1, e2);
                if *is_paren { write!(f, "({})", s) } else { write!(f, "{}", s) }
            },
            Expr::LetRec(func, param, body, cont, is_paren) => {
                let s = format!("let rec {} = fun {} -> {} in {}", func, param, body, cont);
                if *is_paren { write!(f, "({})", s) } else { write!(f, "{}", s) }
            },
            Expr::If(c, t, e, is_paren) => {
                let s = format!("if {} then {} else {}", c, t, e);
                if *is_paren { write!(f, "({})", s) } else { write!(f, "{}", s) }
            },
            Expr::BinOp(e1, op, e2, is_paren) => {
                let s = format!("{} {} {}", e1, op, e2);
                if *is_paren { write!(f, "({})", s) } else { write!(f, "{}", s) }
            }
            Expr::Fun(p, b, is_paren) => {
                let s = format!("fun {} -> {}", p, b);
                if *is_paren { write!(f, "({})", s) } else { write!(f, "{}", s) }
            },
            Expr::App(e1, e2, is_paren) => {
                let s = format!("{} {}", e1, e2);
                if *is_paren { write!(f, "({})", s) } else { write!(f, "{}", s) }
            },
            Expr::Match(e, nil_case, x, y, cons_case, is_paren) => {
                let s = format!("match {} with [] -> {} | {}::{} -> {}", e, nil_case, x, y, cons_case);
                if *is_paren { write!(f, "({})", s) } else { write!(f, "{}", s) }
            },
        }
    }
}

impl fmt::Display for TypeVar {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.name)
    }
}

impl fmt::Display for Type {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Type::Int => write!(f, "int"),
            Type::Bool => write!(f, "bool"),
            Type::Fun(t1, t2) => write!(f, "({} -> {})", t1, t2),
            Type::List(t) => write!(f, "{} list", t),
            Type::Var(tv) => write!(f, "{}", tv),
        }
    }
}

impl fmt::Display for TyScheme {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        if self.vars.is_empty() {
            write!(f, "{}", self.ty)
        } else {
            let mut sorted_vars = self.vars.clone();
            // Sort by the string name to ensure 'a' comes before 'b'.
            sorted_vars.sort_by(|a, b| a.name.cmp(&b.name));

            let vars_str = sorted_vars.iter().map(|v| v.to_string()).collect::<Vec<_>>().join(" ");
            write!(f, "{}. {}", vars_str, self.ty)
        }
    }
}

// --- Helper methods for finding free type variables ---
impl Type {
    pub fn free_type_vars(&self) -> HashSet<TypeVar> {
        let mut ftv = HashSet::new();
        self.collect_ftv(&mut ftv);
        ftv
    }

    fn collect_ftv(&self, ftv: &mut HashSet<TypeVar>) {
        match self {
            Type::Var(tv) => { ftv.insert(tv.clone()); }
            Type::Fun(t1, t2) => {
                t1.collect_ftv(ftv);
                t2.collect_ftv(ftv);
            }
            Type::List(t) => t.collect_ftv(ftv),
            _ => {}
        }
    }
}