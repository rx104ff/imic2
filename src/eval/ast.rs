// src/ast.rs
use std::hash::Hash;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum LanguageVersion {
    ML1,
    ML2,
    ML3,
    ML4,
}
