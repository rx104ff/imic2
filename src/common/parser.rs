use crate::common::{ast::{Expr, Op, Type, Var}, tokenizer::Token};

pub struct ParserCore {
    tokens: Vec<Token>,
    pos: usize,
}

impl ParserCore {
    pub fn new(tokens: Vec<Token>) -> Self {
        ParserCore { tokens, pos: 0 }
    }

    pub fn pos(&self) -> usize {
        self.pos
    }

    pub fn initial(&mut self, initial_pos: usize) {
        self.pos = initial_pos
    }

    pub fn peek(&self) -> Option<&Token> { 
        self.tokens.get(self.pos) 
    }

    pub fn advance(&mut self) {
        if self.pos < self.tokens.len() {
            self.pos += 1;
        }
    }

    pub fn expect(&mut self, expected: Token) -> Result<(), String> {
        match self.peek() {
            Some(token) if *token == expected => { self.advance(); Ok(()) }
            Some(token) => Err(format!("Expected token {:?}, found {:?}", expected, token)),
            None => Err(format!("Expected token {:?}, but found end of input.", expected)),
        }
    }

    pub fn next_var(&mut self) -> Result<Var, String> {
        match self.peek().cloned() {
            Some(Token::Ident(name)) => {
                self.advance();
                Ok(Var(name))
            }
            _ => Err("Expected an identifier.".to_string()),
        }
    }
}

/// A helper function to mark an expression as having been parsed inside parentheses.
fn mark_expr_paren(expr: Expr) -> Expr {
    match expr {
        Expr::Let(v, e1, e2, _) => Expr::Let(v, e1, e2, true),
        Expr::LetRec(f, p, e1, e2, _) => Expr::LetRec(f, p, e1, e2, true),
        Expr::Fun(p, b, _) => Expr::Fun(p, b, true),
        Expr::App(e1, e2, _) => Expr::App(e1, e2, true),
        Expr::If(c, t, e, _) => Expr::If(c, t, e, true),
        Expr::BinOp(e1, op, e2, _) => Expr::BinOp(e1, op, e2, true),
        Expr::Match(e, n, h, t, c, _) => Expr::Match(e, n, h, t, c, true),
        other => other,
    }
}


/// A trait that provides default implementations for parsing all common expressions.
pub trait ExpressionParser {
    // Each implementor must provide access to its core and a way to get a variable.
    fn core(&mut self) -> &mut ParserCore;
    
    // All expression parsing methods are now default methods on this trait.
    // They operate on the required `core` method.
    fn parse_expr(&mut self) -> Result<Expr, String> {
        match self.core().peek() {
            Some(Token::Let) => self.parse_let_expr(),
            Some(Token::If) => self.parse_if_expr(),
            Some(Token::Fun) => self.parse_fun_expr(),
            Some(Token::Match) => self.parse_match_expr(),
            _ => self.parse_lt(),
        }
    }
    
    fn parse_let_expr(&mut self) -> Result<Expr, String> {
        self.core().advance(); // consume 'let'
        if self.core().peek() == Some(&Token::Rec) {
            self.core().advance(); // consume 'rec'
            let func_name = self.core().next_var()?;
            self.core().expect(Token::Equals)?;
            let body = self.parse_expr()?;
            self.core().expect(Token::In)?;
            let cont = self.parse_expr()?;
            if let Expr::Fun(param, fun_body, _) = body {
                 Ok(Expr::LetRec(func_name, param, fun_body, Box::new(cont), false))
            } else { Err("Expected a function definition after 'let rec ='".to_string()) }
        } else {
            let var = self.core().next_var()?;
            self.core().expect(Token::Equals)?;
            let bound_expr = self.parse_expr()?;
            self.core().expect(Token::In)?;
            let cont = self.parse_expr()?;
            Ok(Expr::Let(var, Box::new(bound_expr), Box::new(cont), false))
        }
    }

    fn parse_if_expr(&mut self) -> Result<Expr, String> {
        self.core().advance(); // consume 'if'
        let cond = self.parse_expr()?;
        self.core().expect(Token::Then)?;
        let then_branch = self.parse_expr()?;
        self.core().expect(Token::Else)?;
        let else_branch = self.parse_expr()?;
        Ok(Expr::If(Box::new(cond), Box::new(then_branch), Box::new(else_branch), false))
    }

    fn parse_fun_expr(&mut self) -> Result<Expr, String> {
        self.core().advance(); // consume 'fun'
        let param = self.core().next_var()?;
        self.core().expect(Token::Arrow)?;
        let body = self.parse_expr()?;
        Ok(Expr::Fun(param, Box::new(body), false))
    }

    fn parse_match_expr(&mut self) -> Result<Expr, String> {
        self.core().advance(); // consume 'match'
        let expr_to_match = self.parse_expr()?;
        self.core().expect(Token::With)?;
        self.core().expect(Token::Nil)?;
        self.core().expect(Token::Arrow)?;
        let nil_case = self.parse_expr()?;
        self.core().expect(Token::Bar)?;
        let head_var = self.core().next_var()?;
        self.core().expect(Token::ColonColon)?;
        let tail_var = self.core().next_var()?;
        self.core().expect(Token::Arrow)?;
        let cons_case = self.parse_expr()?;
        Ok(Expr::Match(Box::new(expr_to_match), Box::new(nil_case), head_var, tail_var, Box::new(cons_case), false))
    }

    fn parse_lt(&mut self) -> Result<Expr, String> {
        let mut lhs = self.parse_cons()?;
        while self.core().peek() == Some(&Token::Lt) {
            self.core().advance();
            let rhs = self.parse_cons()?;
            lhs = Expr::BinOp(Box::new(lhs), Op::Lt, Box::new(rhs), false);
        }
        Ok(lhs)
    }
    
    fn parse_cons(&mut self) -> Result<Expr, String> {
        let mut lhs = self.parse_add_sub()?;
        if self.core().peek() == Some(&Token::ColonColon) {
            self.core().advance();
            let rhs = self.parse_cons()?;
            lhs = Expr::BinOp(Box::new(lhs), Op::Cons, Box::new(rhs), false);
        }
        Ok(lhs)
    }

    fn parse_add_sub(&mut self) -> Result<Expr, String> {
        let mut lhs = self.parse_mul()?;
        while let Some(op_token) = self.core().peek() {
            let op = match op_token {
                Token::Plus => Op::Add, Token::Minus => Op::Sub, _ => break,
            };
            self.core().advance();
            let rhs = self.parse_mul()?;
            lhs = Expr::BinOp(Box::new(lhs), op, Box::new(rhs), false);
        }
        Ok(lhs)
    }

    fn parse_mul(&mut self) -> Result<Expr, String> {
        let mut lhs = self.parse_app()?;
        while self.core().peek() == Some(&Token::Star) {
            self.core().advance();
            let rhs = self.parse_app()?;
            lhs = Expr::BinOp(Box::new(lhs), Op::Mul, Box::new(rhs), false);
        }
        Ok(lhs)
    }

    fn parse_app(&mut self) -> Result<Expr, String> {
        let mut lhs = self.parse_atom()?;
        while let Some(tok) = self.core().peek() {
            match tok {
                Token::Int(_) | Token::Bool(_) | Token::Ident(_) | Token::LParen | Token::Nil => {
                    let rhs = self.parse_atom()?;
                    lhs = Expr::App(Box::new(lhs), Box::new(rhs), false);
                },
                _ => break,
            }
        }
        Ok(lhs)
    }

    fn parse_atom(&mut self) -> Result<Expr, String> {
        match self.core().peek().cloned() {
            Some(Token::Int(n)) => { self.core().advance(); Ok(Expr::Int(n)) }
            Some(Token::Bool(b)) => { self.core().advance(); Ok(Expr::Bool(b)) }
            Some(Token::Ident(name)) => { self.core().advance(); Ok(Expr::Var(Var(name))) }
            Some(Token::Nil) => { self.core().advance(); Ok(Expr::Nil) }
            Some(Token::LParen) => {
                self.core().advance();
                let expr = self.parse_expr()?;
                self.core().expect(Token::RParen)?;
                Ok(mark_expr_paren(expr))
            }
            _ => Err("Unexpected token found at atomic level.".to_string())
        }
    }
}

/// A trait for parsers that need to handle type syntax.
pub trait TypeParser {
    // Each implementor must provide access to its core and a way to handle type variables.
    fn core(&mut self) -> &mut ParserCore;
    fn parse_single_type(&mut self) -> Result<Type, String>;

    /// Parses a potentially complex type, like `int -> int` or `(int -> int) list`.
    fn parse_type(&mut self) -> Result<Type, String> {
        let mut ty = self.parse_single_type()?;
        if self.core().peek() == Some(&Token::Arrow) {
            self.core().advance();
            let return_ty = self.parse_type()?;
            ty = Type::Fun(Box::new(ty), Box::new(return_ty));
        }
        Ok(ty)
    }
}