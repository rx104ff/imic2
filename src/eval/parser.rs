// src/parser.rs

use crate::eval::ast::{Expr, Op, Var, Value, Env};
use crate::eval::token::Token;
use std::rc::Rc;

fn mark_expr_paren(expr: Expr) -> Expr {
    match expr {
        Expr::BinOp(e1, op, e2, _) => Expr::BinOp(e1, op, e2, true),
        Expr::App(f, arg, _) => Expr::App(f, arg, true),
        Expr::Fun(param, body, _) => Expr::Fun(param, body, true),
        Expr::Let(v, e1, e2, _) => Expr::Let(v, e1, e2, true),
        Expr::LetRec(f, x, body, e2, _) => Expr::LetRec(f, x, body, e2, true),
        Expr::If(c, t, e, _) => Expr::If(c, t, e, true),
        Expr::Match(e, nil_case, hd, tl, cons_case, _) => Expr::Match(e, nil_case, hd, tl, cons_case, true),
        Expr::Cons(h, t, _) => Expr::Cons(h, t, true),
        other => other,
    }
}

pub struct Parser {
    tokens: Vec<Token>,
    pos: usize,
}

impl Parser {
    pub fn new(tokens: Vec<Token>) -> Self {
        Parser { tokens, pos: 0 }
    }

    fn peek(&self) -> Option<&Token> {
        self.tokens.get(self.pos)
    }

    fn advance(&mut self) {
        self.pos += 1;
    }

    fn expect(&mut self, expected: &Token) {
        let token = self.peek().cloned();
        if token.as_ref() != Some(expected) {
            panic!("Expected '{:?}', found '{:?}'", expected, token);
        }
        self.advance();
    }

    fn next_var(&mut self) -> Var {
        let token = self.peek().cloned();
        match token {
            Some(Token::Ident(name)) => {
                self.advance();
                Var::from(name)
            }
            _ => panic!("Expected identifier, found '{:?}'", token),
        }
    }

    pub fn parse_program(&mut self) -> (Env, Expr) {
        let env = if let Some(Token::Ident(_)) = self.peek() {
            self.parse_env_list()
        } else {
            Rc::new(vec![])
        };
        self.expect(&Token::Bar);
        self.expect(&Token::Minus);
        let expr = self.parse_expr();
        if let Some(Token::Evalto) = self.peek() {
            while self.peek().is_some() && self.peek() != Some(&Token::EOF) {
                self.advance();
            }
        }
        (env, expr)
    }

    fn parse_env_list(&mut self) -> Env {
        let mut bindings = vec![];
        loop {
            let token = self.peek().cloned();
            match token {
                Some(Token::Ident(_)) => {
                    let var = self.next_var();
                    self.expect(&Token::Equals);
                    let val = self.parse_value();
                    bindings.push((var, val));
                    if matches!(self.peek(), Some(Token::Comma)) {
                        self.advance();
                    } else {
                        break;
                    }
                }
                _ => break,
            }
        }
        Rc::new(bindings)
    }

    fn parse_value(&mut self) -> Value {
    self.parse_list_value(false)
}

fn parse_list_value(&mut self, paren: bool) -> Value {
    let left = self.parse_single_value();
    self.parse_list_tail(left, paren)
}

fn parse_list_tail(&mut self, left: Value, paren: bool) -> Value {
    if let Some(Token::ColonColon) = self.peek() {
        self.advance();
        let right = self.parse_list_value(false);
        Value::Cons(Box::new(left), Box::new(right), paren)
    } else {
        left
    }
}

    fn parse_single_value(&mut self) -> Value {
        match self.peek().cloned() {
            Some(Token::Int(n)) => {
                self.advance();
                Value::Int(n)
            }
            Some(Token::Bool(b)) => {
                self.advance();
                Value::Bool(b)
            }
            Some(Token::LParen) => {
                self.advance();
                let saved_pos = self.pos;
                let env = self.parse_env_list();

                // Try to parse as a function environment
                if self.peek() == Some(&Token::RParen) {
                    self.advance();
                    if self.peek() == Some(&Token::LBracket) {
                        self.advance();
                        match self.peek() {
                            Some(Token::Fun) => return self.parse_func_val(env),
                            Some(Token::Rec) => return self.parse_rec_func_val(env),
                            _ => panic!("Expected fun or rec after ["),
                        }
                    }
                }

                // Not a function env, so rewind and try to parse as a value in parens
                self.pos = saved_pos;
                let inner = self.parse_value();
                self.expect(&Token::RParen);

                match inner {
                    Value::Cons(l, r, _) => Value::Cons(l, r, true),
                    _ => inner,
                }
            }
            Some(Token::LBracket) => {
                self.advance();
                self.expect(&Token::RBracket);
                Value::Nil
            }
            _ => panic!("Invalid value syntax: {:?}", self.peek()),
        }
    }

    fn collect_tokens_until_rbracket(&mut self) -> Vec<Token> {
        let mut tokens = vec![];
        let mut depth = 1;
        while let Some(token) = self.peek().cloned() {
            self.advance();
            match token {
                Token::LBracket => {
                    depth += 1;
                }
                Token::RBracket => {
                    depth -= 1;
                    if depth == 0 {
                        break;
                    }
                }
                _ => {}
            }
            tokens.push(token);
        }
        tokens
    }

    fn parse_func_val(&mut self, env: Env) -> Value {
        self.expect(&Token::Fun);
        let param = self.next_var();
        self.expect(&Token::Arrow);
        let body_tokens = self.collect_tokens_until_rbracket();
        let mut inner_parser = Parser::new(body_tokens);
        let body = inner_parser.parse_expr();
        Value::FunVal(param, Box::new(body), env, false)
    }

    fn parse_rec_func_val(&mut self, env: Env) -> Value {
        self.expect(&Token::Rec);
        let name: Var = self.next_var();
        self.expect(&Token::Equals);
        self.expect(&Token::Fun);
        let param = self.next_var();
        self.expect(&Token::Arrow);
        let body_tokens = self.collect_tokens_until_rbracket();
        let mut inner_parser = Parser::new(body_tokens);
        let body = inner_parser.parse_expr();
        Value::RecFunVal(name, param, Box::new(body), env, false)
    }

    fn parse_expr(&mut self) -> Expr {
        self.parse_lt()
    }

    fn parse_lt(&mut self) -> Expr {
        let mut expr = self.parse_cons(); // parse lhs first
        while let Some(Token::Lt) = self.peek() {
            self.advance();
            let right = self.parse_cons();
            expr = Expr::BinOp(Box::new(expr), Op::Lt, Box::new(right), false);
        }
        expr
    }

    fn parse_cons(&mut self) -> Expr {
        self.parse_cons_prec(false)
    }

    fn parse_cons_prec(&mut self, paren: bool) -> Expr {
        let lhs = self.parse_add_sub_prec(false);
        self.parse_cons_tail(lhs, paren)
    }

    fn parse_cons_tail(&mut self, lhs: Expr, paren: bool) -> Expr {
        if let Some(Token::ColonColon) = self.peek() {
            self.advance();
            let rhs = self.parse_cons_prec(false);
            Expr::BinOp(Box::new(lhs), Op::Cons, Box::new(rhs), paren)
        } else {
            lhs
        }
    }

    fn parse_add_sub_prec(&mut self, paren: bool) -> Expr {
        let mut expr = self.parse_mul_div_prec(false);
        while let Some(op_token) = self.peek() {
            let op = match op_token {
                Token::Plus => Op::Add,
                Token::Minus => Op::Sub,
                _ => break,
            };
            self.advance();
            let right = self.parse_mul_div_prec(false);
            expr = Expr::BinOp(Box::new(expr), op, Box::new(right), paren);
        }
        expr
    }

    fn parse_mul_div_prec(&mut self, paren: bool) -> Expr {
        let mut expr = self.parse_app_or_atom();
        while let Some(op_token) = self.peek() {
            let op = match op_token {
                Token::Star => Op::Mul,
                _ => break,
            };
            self.advance();
            let right = self.parse_app_or_atom();
            expr = Expr::BinOp(Box::new(expr), op, Box::new(right), paren);
        }
        expr
    }

    fn parse_app_or_atom(&mut self) -> Expr {
        let mut expr = self.parse_atom();
        while matches!(self.peek(), Some(Token::Ident(_)) | Some(Token::Int(_)) | Some(Token::LParen) | Some(Token::Nil)) {
            let arg = self.parse_atom();
            expr = Expr::App(Box::new(expr), Box::new(arg), false);
        }
        expr
    }

    fn parse_atom(&mut self) -> Expr {
        match self.peek() {
            Some(Token::Minus) => {
                self.advance();
                match self.peek().cloned() {
                    Some(Token::Int(n)) => {
                        self.advance();
                        Expr::Int(-(n))
                    }
                    _ => {
                        // Parse as unary minus: `-expr` => `0 - expr`
                        let expr = self.parse_atom();
                        Expr::BinOp(Box::new(Expr::Int(0)), Op::Sub, Box::new(expr), false)
                    }
                }
            }
            Some(Token::If) => self.parse_if(),
            Some(Token::Let) => self.parse_let(),
            Some(Token::Ident(_)) => Expr::Var(self.next_var()),
            Some(Token::Int(_)) => self.parse_int(),
            Some(Token::LParen) => {
                self.advance();
                let expr = self.parse_expr();
                self.expect(&Token::RParen);
                mark_expr_paren(expr)
            }
            Some(Token::Fun) => self.parse_fun_expr(),
            Some(Token::Nil) => {
                self.advance();
                Expr::Nil
            },
            Some(Token::Match) => self.parse_match(),
            _ => panic!("Unexpected token in expression: {:?}", self.peek()),
        }
    }

    fn parse_if(&mut self) -> Expr {
        self.expect(&Token::If);
        let cond = self.parse_expr();
        self.expect(&Token::Then);
        let then_branch = self.parse_expr();
        self.expect(&Token::Else);
        let else_branch = self.parse_expr();
        Expr::If(Box::new(cond), Box::new(then_branch), Box::new(else_branch), false)
    }

    fn parse_let(&mut self) -> Expr {
        self.expect(&Token::Let);
        if self.peek() == Some(&Token::Rec) {
            self.advance(); // consume 'rec'
            let name    = self.next_var(); // 'name'
            self.expect(&Token::Equals); // '='
            self.expect(&Token::Fun); // 'fun'
            let param = self.next_var(); // 'n'
            self.expect(&Token::Arrow); // '->'
            let body = self.parse_expr(); // parse the function body
            self.expect(&Token::In); // 'in'
            let cont = self.parse_expr(); // continuation expression
            Expr::LetRec(name, param, Box::new(body), Box::new(cont), false)
        } else {
            let name = self.next_var();
            self.expect(&Token::Equals);
            let bound_expr = self.parse_expr();
            self.expect(&Token::In);
            let cont = self.parse_expr();
            Expr::Let(name, Box::new(bound_expr), Box::new(cont), false)
        }
    }

    fn parse_int(&mut self) -> Expr {
        let token = self.peek().cloned();
        match token {
            Some(Token::Int(n)) => {
                self.advance();
                Expr::Int(n)
            }
            _ => panic!("Expected int"),
        }
    }

    fn parse_fun_expr(&mut self) -> Expr {
        self.expect(&Token::Fun);
        let param = self.next_var();
        self.expect(&Token::Arrow);
        let body = self.parse_expr();
        Expr::Fun(param, Box::new(body), false)
    }

    fn parse_match(&mut self) -> Expr {
        self.expect(&Token::Match);
        let expr = self.parse_expr();
        self.expect(&Token::With);

        // Case 1: [] -> e2
        self.expect(&Token::Nil);
        self.expect(&Token::Arrow);
        let nil_case = self.parse_expr();

        // Case 2: | x :: y -> e3
        self.expect(&Token::Bar);
        let hd = self.next_var();
        self.expect(&Token::ColonColon);
        let tl = self.next_var();
        self.expect(&Token::Arrow);
        let cons_case = self.parse_expr();

        Expr::Match(Box::new(expr), Box::new(nil_case), hd, tl, Box::new(cons_case), false)
    }


}
