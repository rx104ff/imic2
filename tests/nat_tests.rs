use imic2::nat;
use imic2::nat::ast::ComparisonMode;
use imic2::nat::parser::Parser;

use insta::assert_snapshot;


fn run_nat_test(mode: Option<ComparisonMode>, input: &str) -> String {
    let tokens = nat::token::tokenize(input);
    let mut parser = Parser::new(tokens);
    let judgment = match parser.parse_judgment() {
        Ok(j) => j,
        Err(e) => return format!("Parsing Error: {}", e),
    };

    match nat::nat::derive(&judgment, mode) {
        Ok(derivation) => format!("{}", derivation),
        Err(e) => format!("Derivation Error: {}", e),
    }
}

#[test]
fn test_nat_1() {
    let result = run_nat_test(None, "Z plus Z is Z");
    assert_snapshot!(result);
}

#[test]
fn test_nat_2() {
    let result = run_nat_test(None, "Z plus S(S(Z)) is S(S(Z))");
    assert_snapshot!(result);
}

#[test]
fn test_nat_3() {
    let result = run_nat_test(None, "S(S(Z)) plus Z is S(S(Z))");
    assert_snapshot!(result);
}

#[test]
fn test_nat_4() {
    let result = run_nat_test(None, "S(Z) plus S(S(S(Z))) is S(S(S(S(Z))))");
    assert_snapshot!(result);
}

#[test]
fn test_nat_5() {
    let result = run_nat_test(None, "Z times S(S(Z)) is Z");
    assert_snapshot!(result);
}

#[test]
fn test_nat_6() {
    let result = run_nat_test(None, "S(S(Z)) times Z is Z");
    assert_snapshot!(result);
}

#[test]
fn test_nat_7() {
    let result = run_nat_test(None, "S(S(Z)) times S(Z) is S(S(Z))");
    assert_snapshot!(result);
}

#[test]
fn test_nat_8() {
    let result = run_nat_test(None, "S(S(Z)) times S(S(Z)) is S(S(S(S(Z))))");
    assert_snapshot!(result);
}

#[test]
fn test_compare_nat_1_9() {
    let result = run_nat_test(Some(ComparisonMode::V1), "S(S(Z)) is less than S(S(S(Z)))");
    assert_snapshot!(result);
}

#[test]
fn test_compare_nat_2_10() {
    let result = run_nat_test(Some(ComparisonMode::V2), "S(S(Z)) is less than S(S(S(Z)))");
    assert_snapshot!(result);
}

#[test]
fn test_compare_nat_3_11() {
    let result = run_nat_test(Some(ComparisonMode::V3), "S(S(Z)) is less than S(S(S(Z)))");
    assert_snapshot!(result);
}
// #[test]
// fn test_compare_v2_success() {
//     let result = run_nat_test(Some(ComparisonMode::V2), "S(Z) is less than S(S(Z))");
//     assert_snapshot!(result);
// }

// #[test]
// fn test_invalid_judgment_error() {
//     let result = run_nat_test(Some(ComparisonMode::V1), "S(Z) is less than Z");
//     // You can snapshot error messages too!
//     assert_snapshot!(result);
// }