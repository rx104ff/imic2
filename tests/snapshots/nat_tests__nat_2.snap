---
source: tests/nat_tests.rs
expression: result
---
Z plus S(S(Z)) is S(S(Z)) by P-Zero {
};
